File: db/schema/personal-imports.ts
import {
  pgTable,
  uuid,
  text,
  integer,
  timestamp,
  varchar,
  index,
} from 'drizzle-orm/pg-core';
import { personalAccounts } from './accounts';

/**
 * One row per CSV statement upload. Tracks the S3 object, parser
 * choice, and per-import counts so the user can see what happened on
 * the Imports page.
 *
 * Lifecycle:
 *   pending     row created, file uploaded to S3 (PUT in flight)
 *   processing  S3 event fired, CSV parser running
 *   ready       parse succeeded — see `imported_count` / `skipped_count`
 *   failed      parse errored — see `error`
 */
export const personalImports = pgTable(
  'personal_imports',
  {
    id: uuid('id').primaryKey().defaultRandom(),

    account_id: uuid('account_id')
      .notNull()
      .references(() => personalAccounts.id),
    institution: varchar('institution', { length: 30 }).notNull(),
    //   Drives parser dispatch. Stored on the import row so a future
    //   reparse uses the same parser the user picked at upload time.

    status: varchar('status', { length: 20 }).notNull().default('pending'),

    s3_key: text('s3_key').notNull(),
    filename: text('filename'),
    size_bytes: integer('size_bytes'),

    row_count: integer('row_count'),         // total rows in the file
    imported_count: integer('imported_count'), // newly inserted
    skipped_count: integer('skipped_count'), // duplicates skipped

    error: text('error'),

    createdAt: timestamp('created_at', { withTimezone: true })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp('updated_at', { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (t) => [
    index('personal_imports_account_idx').on(t.account_id),
    index('personal_imports_status_idx').on(t.status),
  ],
);

export type PersonalImport = typeof personalImports.$inferSelect;
export type NewPersonalImport = typeof personalImports.$inferInsert;

File: services/api/app/models/personal_imports.py
"""Pydantic models for the ``personal_imports`` table."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.models.personal_accounts import Institution


class PersonalImportRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    account_id: UUID
    institution: Institution
    status: str
    s3_key: str
    filename: str | None = None
    size_bytes: int | None = None
    row_count: int | None = None
    imported_count: int | None = None
    skipped_count: int | None = None
    error: str | None = None
    created_at: datetime
    updated_at: datetime


class PersonalImportCreateRequest(BaseModel):
    """``POST /personal/imports/`` body.

    The client picks which account this CSV belongs to and which
    parser to use. We sign a presigned PUT for the file body and
    return both the new row + the upload URL.
    """

    account_id: UUID
    institution: Institution
    filename: str | None = None
    size_bytes: int | None = None


class PersonalImportCreateResponse(BaseModel):
    import_: PersonalImportRead
    upload_url: str

    model_config = ConfigDict(populate_by_name=True)
